errors/error.go

```diff
@@ -0,0 +1,373 @@
+// It's heavily influenced by Rob Pike `errors` package in the Upspin project:
+// 	https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
+// See E() for its usage.
+type Error struct {
+	// Kind is the kind of error.
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack StackMeta
+
+	// Err represents the underlying error.
+	Err error
+}
+
+type (
+	// Kind defines the kind of error.
+	Kind string
+	// StackMeta has the metadata of the stack which originated the error.
+	// Same interface as stack.Metadata.
+	StackMeta interface {
+		Name() string
+		Desc() string
+		Path() string
+	}
+)
+
+const separator = ": "
+
+// E builds an error value from its arguments.
+// There must be at least one argument or E panics.
+// The type of each argument determines its meaning. If more than one argument
+// of a given type is presented, only the last one is recorded.
+//
+// The types are:
+// 	errors.Kind
+//		The kind of error (eg.: HCLSyntax, TerramateSchema, etc).
+//	hcl.Range
+//		The file range where the error originated.
+//  errors.Stack
+//		The stack which the error originated.
+//	error
+//		The underlying error that triggered this one.
+//	hcl.Diagnostics
+//		The underlying hcl error that triggered this one.
+//		If this error's FileRange is not set, the diagnostic subject range is
+//		pulled.
+//		If this error's Description is not set, the diagnostic detail field is
+//		pulled.
+//	string
+//		The error description. It supports formatting using the Go's fmt verbs
+//      as long as the arguments are not of the defined types.
+//
+// If the error is printed, only those items that have been
+// set to non-zero values will appear in the result. For the `hcl.Range` type,
+// the `range.Empty()` method is used.
+//
+// Error promotions:
+//
+// If Kind is not specified or Any, we set it to the Kind of
+// the underlying error.
+// If FileRange is not specified, we set it to the FileRange of the underlying
+// error (if set).
+//
+// Minimization:
+//
+// In order to avoid duplicated messages, we erase the fields present in the
+// underlying error if already set with same value in this error.
+func E(args ...interface{}) error {
+	if len(args) == 0 {
+		panic("called with no args")
+	}
+
+	var (
+		diags  hcl.Diagnostics
+		format *string
+	)
+
+	fmtargs := []interface{}{}
+
+	e := &Error{}
+	for _, arg := range args {
+		switch arg := arg.(type) {
+		case Kind:
+			e.Kind = arg
+		case hcl.Range:
+			e.FileRange = arg
+		case hcl.Diagnostics:
+			diags = arg
+			if diags.HasErrors() {
+				diag := diags[0]
+				if diag.Subject != nil {
+					e.FileRange = *diag.Subject
```

Comment:
The range...on the diagnostics...is called SUBJECT ?

![terror](https://raw.githubusercontent.com/katcipis/memes/master/tom-shock.jpg)

Reply:
it kinda makes sense for me... because just Range says nothing, FileRange says a bit more but still what range? The word *Subject* at least points that this field contains the offending part, the topic of the diagnostics, and so on. They also provide a *Context* field with a broader range so you can print/render this part of the code in an error in order to easily spot the problem because sometimes the `Subject` is a tiny expression inside an expression and so on.
But Context can be nil, which means the Context == Subject, but they don't provide any helper method for that. For the extension the tighter range is enough but I think we can provide better error messages using the Context (if we get rid of logging for showing errors in text mode).

Reply:
was unaware they had two ranges, if there is just one it is simple to me that it is the file range of the source that caused the problem/diagnostic and calling subject is just confusing, but if they have multiple ranges then it may make sense.